Add nrf54l09 EngA SoC and nrf54l09 PDK

nordic-krch opened the pull request:
Add SoC and PDK board.

nordic-krch committed:
soc: nordic: nrf54l: Add nrf54l09 enga SoC

Add nrf54l09 EngA SoC in soc, dts and hal_nordic.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
boards: nordic: Add nrf54l09pdk

Add nrf54l09pdk board with nrf54l09 EngA SoC.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

masz-nordic commented on soc/nordic/soc.yml:
```diff
@@ -25,6 +25,10 @@ family:
+          - name: nrf54l09
+            cpuclusters:
+              - name: cpuapp
+              - name: cpuflpr
```
FLPR is not added yet, so it should be skipped here for now.

masz-nordic commented on modules/hal_nordic/nrfx/nrfx_config.h:
```diff
@@ -1133,6 +1133,10 @@
     #include <nrfx_config_nrf54l05_flpr.h>
+#elif (defined(NRF54L09_XXAA) | defined(NRF54L09_ENGA_XXAA)) && defined(NRF_APPLICATION)
+    #include <nrfx_config_nrf54l09_enga_application.h>
+#elif defined(NRF54L09_XXAA) && defined(NRF_FLPR)
+    #include <nrfx_config_nrf54l09_flpr.h>
```
Same with FLPR.

masz-nordic commented on modules/hal_nordic/nrfx/CMakeLists.txt:
```diff
@@ -232,6 +234,7 @@ mdk_svd_ifdef(CONFIG_SOC_NRF54L05_CPUAPP      nrf54l05_application.svd)
 mdk_svd_ifdef(CONFIG_SOC_NRF54L05_CPUFLPR     nrf54l05_flpr.svd)
 mdk_svd_ifdef(CONFIG_SOC_NRF54L10_CPUAPP      nrf54l10_application.svd)
 mdk_svd_ifdef(CONFIG_SOC_NRF54L10_CPUFLPR     nrf54l10_flpr.svd)
+mdk_svd_ifdef(CONFIG_SOC_NRF54L09_ENGA_CPUAPP nrf54l09_enga_application.svd)
```
Should be before L10

masz-nordic replied:
Not fixed?

nordicjm commented on boards/nordic/nrf54l09pdk/doc/index.rst:
```diff
@@ -0,0 +1,91 @@
+Supported Features
+==================
+
+The ``nrf54l09pdk/nrf54l09/cpuapp`` board target configuration supports the following
```
-configuration

nordicjm commented on boards/nordic/nrf54l09pdk/nrf54l09pdk_nrf54l09-common.dtsi:
```diff
@@ -0,0 +1,72 @@
+/ {
+	leds {
+		compatible = "gpio-leds";
+		/*led0: led_0 {*/
+			/*gpios = <&gpio2 10 GPIO_ACTIVE_HIGH>;*/
+			/*label = "Green LED 0";*/
+		/*};*/
```
remove if not needed

nordicjm commented on boards/nordic/nrf54l09pdk/nrf54l09pdk_nrf54l09-common.dtsi:
```diff
@@ -0,0 +1,72 @@
+	};
+
+	aliases {
+		/*led0 = &led0;*/
```
as above

nordic-krch committed:
tests: lib: cpp: cxx: Filter out all ARMs from nRF54x and nRF92

Instead of extending list of unsupported platforms use filter that
will exclude or current and next ARM cores from platforms with HAL
that does not support C++98.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

e-rk commented on dts/arm/nordic/nrf54l09_enga_cpuapp.dtsi:
```diff
@@ -0,0 +1,8 @@
+ */
+
+#include <nordic/nrf54l09.dtsi>
+#include "nrf54l_05_10_15_cpuapp.dtsi"
```
I have a PR that enables `dppic` and `ppib` nodes that should be merged today: https://github.com/zephyrproject-rtos/zephyr/pull/82254

Could you update this file and add the same DTS overrides here?

nordic-krch replied:
it's including `nrf54l_05_10_15_cpuapp.dtsi` which #82254 updates. Do i need to add it something on top of that?

e-rk replied:
Ah, I see. In that case nothing more is needed.

e-rk commented on boards/nordic/nrf54l09pdk/nrf54l09_cpuapp_common.dtsi:
```diff
@@ -0,0 +1,113 @@
+		slot1_ns_partition: partition@a6000 {
+			label = "image-1-nonsecure";
+			reg = <0xa6000 DT_SIZE_K(200)>;
+		};
```
Looks like the `storage_partition` is missing.
In nrf54l15 there is this:
```
		storage_partition: partition@15c000 {
			label = "storage";
			reg = <0x15c000 DT_SIZE_K(36)>;
		};
```

nordic-krch replied:
done

e-rk commented:
I have observed that when `CONFIG_CLOCK_CONTROL_NRF=y` the SoC gets stuck at init

```
__set_BASEPRI (basePri=32) at /home/rak1/git/zephyr/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:1315
1315	  __ASM volatile ("MSR basepri, %0" : : "r" (basePri) : "memory");
(gdb) bt
#0  __set_BASEPRI (basePri=32) at /home/rak1/git/zephyr/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:1315
#1  arch_irq_unlock (key=32) at /home/rak1/git/zephyr/zephyr/include/zephyr/arch/arm/asm_inline_gcc.h:89
#2  arch_cpu_atomic_idle (key=key@entry=32) at /home/rak1/git/zephyr/zephyr/arch/arm/core/cortex_m/cpu_idle.c:136
#3  0x00002f3e in k_cpu_atomic_idle (key=32) at /home/rak1/git/zephyr/zephyr/include/zephyr/kernel.h:6023
#4  lfclk_spinwait (mode=CLOCK_CONTROL_NRF_LF_START_STABLE)
    at /home/rak1/git/zephyr/zephyr/drivers/clock_control/clock_control_nrf.c:515
#5  z_nrf_clock_control_lf_on (start_mode=start_mode@entry=CLOCK_CONTROL_NRF_LF_START_STABLE)
    at /home/rak1/git/zephyr/zephyr/drivers/clock_control/clock_control_nrf.c:571
#6  0x000036c6 in sys_clock_driver_init () at /home/rak1/git/zephyr/zephyr/drivers/timer/nrf_grtc_timer.c:505
#7  sys_clock_driver_init () at /home/rak1/git/zephyr/zephyr/drivers/timer/nrf_grtc_timer.c:468
#8  0x00004dae in z_sys_init_run_level (level=level@entry=INIT_LEVEL_PRE_KERNEL_2)
    at /home/rak1/git/zephyr/zephyr/kernel/init.c:371
#9  0x00004fbc in z_cstart () at /home/rak1/git/zephyr/zephyr/kernel/init.c:779
#10 0x00001d28 in z_prep_c () at /home/rak1/git/zephyr/zephyr/arch/arm/core/cortex_m/prep_c.c:209
#11 0x00001b56 in z_arm_reset () at /home/rak1/git/zephyr/zephyr/arch/arm/core/cortex_m/reset.S:169
```

e-rk commented on dts/common/nordic/nrf54l09.dtsi:
```diff
@@ -0,0 +1,60 @@
+};
+
+&cpuapp_rram {
+	reg = <0x0 DT_SIZE_K(950)>;
```
There is a problem with sizes that are not a multiple of `erase-block-size`.
https://github.com/zephyrproject-rtos/zephyr/blob/bd69ae46bc63821d0eb0955611e4804ed7d14f16/drivers/flash/soc_flash_nrf_rram.c#L52
The number of pages will be rounded-down when using the `nrf_rram_page_layout`.

nordic-krch replied:
changed to 952 and 60 for flpr.

e-rk replied:
Was this change commited? It seems the value is still at `950`.

nordic-krch replied:
not sure what happen but should be ok now.

e-rk commented on dts/common/nordic/nrf54l09.dtsi:
```diff
@@ -0,0 +1,60 @@
+
+/* 952 + 60 = 1012KB */
+&rram_controller {
+	cpuflpr_rram: rram@ed800 {
```
```suggestion
	cpuflpr_rram: rram@ee000 {
```

nordic-krch replied:
done

nordicjm commented:
rebase needed

e-rk commented:
> I have observed that when `CONFIG_CLOCK_CONTROL_NRF=y` the SoC gets stuck at init
> 
> ...

This needs to be added to `nrf54l09.dtsi`
```
&clock {
	interrupts = <270 NRF_DEFAULT_IRQ_PRIORITY>;
};
```

masz-nordic commented on dts/common/nordic/nrf54l09.dtsi:
```diff
@@ -0,0 +1,68 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "nrf54l_05_10_15.dtsi"
+
+/delete-node/ &gpio2;
+/delete-node/ &spi00;
+/delete-node/ &uart00;
+/delete-node/ &spi22;
+/delete-node/ &i2c22;
+/delete-node/ &uart22;
+/delete-node/ &pdm20;
+/delete-node/ &pdm21;
+/delete-node/ &pwm20;
+/delete-node/ &pwm21;
+/delete-node/ &pwm22;
+/delete-node/ &i2s20;
+/delete-node/ &qdec20;
+/delete-node/ &qdec21;
+/delete-node/ &nfct;
```
The `nrf54l_05_10_15.dtsi` was added due to the fact that those are identical HW wise, except for memory.
L09 is not, so I think it should not depend on that file.

e-rk replied:
Removed the dependency and defined all nodes from scratch in `nrf54l09.dtsi`

masz-nordic commented on soc/nordic/soc.yml:
```diff
@@ -162,6 +167,8 @@ runners:
           - qualifiers:
               - nrf9161
               - nrf9161/ns
+          - qualifiers:
+              - nrf54l09/cpuapp
```
Should be after 05

e-rk replied:
Reordered

masz-nordic commented on dts/arm/nordic/nrf54l09_enga_cpuapp.dtsi:
```diff
@@ -0,0 +1,8 @@
+ */
+
+#include <nordic/nrf54l09.dtsi>
+#include "nrf54l_05_10_15_cpuapp.dtsi"
```
Either the file should be renamed or separate file for 09 app core should be added.

e-rk replied:
I have inlined the contents from `#include "nrf54l_05_10_15_cpuapp.dtsi"`, similar to how it was done in nrf54l20. Let me know if you want this to be done some other way.

anangl commented on modules/hal_nordic/nrfx/nrfx_config.h:
```diff
@@ -1136,6 +1136,8 @@
     #include <nrfx_config_nrf54l05_application.h>
 #elif defined(NRF54L05_XXAA) && defined(NRF_FLPR)
     #include <nrfx_config_nrf54l05_flpr.h>
+#elif (defined(NRF54L09_XXAA) | defined(NRF54L09_ENGA_XXAA)) && defined(NRF_APPLICATION)
```
What's the point in checking the `NRF54L09_XXAA` symbol here? It doesn't seem to be defined/supported anywhere.

e-rk replied:
Removed the `NRF54L09_XXAA` symbol.

anangl commented on boards/nordic/nrf54l09pdk/nrf54l09pdk_nrf54l09-common.dtsi:
```diff
@@ -0,0 +1,67 @@
+			label = "Green LED 2";
+		};
+		led3: led_3 {
+			gpios = <&gpio1 7 GPIO_ACTIVE_HIGH>;
```
This GPIO is already assigned to `led1` (copy-paste mistake?): https://github.com/zephyrproject-rtos/zephyr/blob/b6e2e113a138344a213ed7b9252b90dd549e0f42/boards/nordic/nrf54l09pdk/nrf54l09pdk_nrf54l09-common.dtsi#L12-L13

e-rk replied:
Removed the duplicated `led`
